File: java/src/main/java/money_problem/domain/Portfolio.java
package money_problem.domain;

import io.vavr.collection.Seq;
import io.vavr.collection.Vector;
import io.vavr.control.Either;

import static io.vavr.control.Either.left;
import static io.vavr.control.Either.right;

public final class Portfolio {
    private final Seq<Money> moneys;

    public Portfolio() {
        this.moneys = Vector.empty();
    }

    private Portfolio(Seq<Money> moneys) {
        this.moneys = moneys;
    }

    public Portfolio add(Money money) {
        return new Portfolio(moneys.append(money));
    }

    private Seq<Either<Error, Money>> convertAllMoneys(Bank bank, Currency toCurrency) {
        return moneys.map(money -> bank.convert(money, toCurrency));
    }

    private boolean containsFailure(Seq<Either<Error, Money>> convertedMoneys) {
        return convertedMoneys.exists(Either::isLeft);
    }

    private Error toFailure(Seq<Either<Error, Money>> convertedMoneys) {
        return new Error(convertedMoneys
                .filter(Either::isLeft)
                .map(e -> String.format("[%s]", e.getLeft().message()))
                .mkString("Missing exchange rate(s): ", ",", ""));
    }

    private Money sumConvertedMoney(Seq<Either<Error, Money>> convertedMoneys, Currency toCurrency) {
        return new Money(convertedMoneys
                .filter(Either::isRight)
                .map(e -> e.getOrElse(new Money(0, toCurrency)))
                .map(Money::amount)
                .reduce(Double::sum), toCurrency);
    }

    public Either<Error, Money> evaluate(Bank bank, Currency to) {
        var convertedMoneys = convertAllMoneys(bank, to);

        return containsFailure(convertedMoneys)
                ? left(toFailure(convertedMoneys))
                : right(sumConvertedMoney(convertedMoneys, to));
    }
}

File: java/src/main/java/money_problem/domain/Portfolio.kt
package money_problem.domain

import io.vavr.control.Either
import io.vavr.control.Either.left
import io.vavr.control.Either.right

class Portfolio {
    private val moneys: List<Money>

    constructor() {
        moneys = emptyList()
    }

    private constructor(moneys: List<Money>) {
        this.moneys = moneys
    }

    fun add(money: Money): Portfolio = Portfolio(moneys + money)

    fun evaluate(bank: Bank, to: Currency): Either<Error, Money> =
        convertAllMoneys(bank, to)
            .let {
                if (containsFailure(it)) left(toFailure(it))
                else right(sumConvertedMoney(it, to))
            }

    private fun convertAllMoneys(bank: Bank, toCurrency: Currency): List<Either<Error, Money>> =
        moneys.map { money: Money -> bank.convert(money, toCurrency) }

    private fun containsFailure(convertedMoneys: List<Either<Error, Money>>): Boolean =
        convertedMoneys.any { result -> result.isLeft }

    private fun toFailure(convertedMoneys: List<Either<Error, Money>>): Error =
        Error(convertedMoneys
            .filter { result -> result.isLeft }
            .joinToString(",", "Missing exchange rate(s): ") { result -> String.format("[%s]", result.left.message) })

    private fun sumConvertedMoney(convertedMoneys: List<Either<Error, Money>>, toCurrency: Currency): Money =
        Money(
            convertedMoneys
                .filter { result -> result.isRight }
                .map { result -> result.getOrElse(Money(0.0, toCurrency)) }
                .sumOf { result -> result.amount },
            toCurrency
        )
}